Mathlib/Analysis/AperiodicOrder/Delone/Basic.lean

```diff
@@ -0,0 +1,172 @@
+def RelativelyDense (D : Set X) : Prop :=
+    ∃ R > 0, ∀ x : X, ∃ y ∈ D, dist x y ≤ R
+
+/-- If `D ⊆ E` and `E` is uniformly discrete, then so is `D`. -/
+lemma uniformlyDiscrete_mono {D E : Set X} (hDE : D ⊆ E) :
+    UniformlyDiscrete E → UniformlyDiscrete D := by
+  rintro ⟨r, hr, hsep⟩
+  refine ⟨r, hr, fun x y hx hy hne ↦ ?_⟩
+  exact hsep (hDE hx) (hDE hy) hne
+
+/-- If `D ⊆ E` and `D` is relatively dense, then so is `E`. -/
+lemma relativelyDense_mono {D E : Set X} (hDE : D ⊆ E) :
+    RelativelyDense D → RelativelyDense E := by
+  rintro ⟨R, hR, hcov⟩
+  refine ⟨R, hR, fun x ↦ ?_⟩
+  rcases hcov x with ⟨y, hyD, hxy⟩
+  exact ⟨y, hDE hyD, hxy⟩
+
+end Metric
+
+namespace Delone
+
+/-- A **Delone set** in a metric space: uniformly discrete and relatively dense. -/
+structure DeloneSet (X : Type*) [MetricSpace X] where
+  /-- The underlying set of a Delone set. -/
+  (carrier : Set X)
+  /-- Uniform discreteness: distinct points of the set are separated by a positive distance. -/
+  (uniformlyDiscrete : UniformlyDiscrete carrier)
+  /-- Relative denseness: every point of the space is within some bounded distance of the set. -/
+  (relativelyDense : RelativelyDense carrier)
+
+attribute [simp] DeloneSet.carrier
+
+namespace DeloneSet
+
+/-- A Delone set is nonempty. -/
+lemma nonempty [Inhabited X] (D : DeloneSet X) : D.carrier.Nonempty := by
+  obtain ⟨R, hR, hcov⟩ := D.relativelyDense
+  rcases hcov (default : X) with ⟨y, hyD, _⟩
+  exact ⟨y, hyD⟩
+
+/-- Extract the covering radius of a Delone set. -/
+lemma exists_covering_radius (D : DeloneSet X) :
+    ∃ R > 0, ∀ x : X, ∃ y ∈ D.carrier, dist x y ≤ R :=
+  D.relativelyDense
+
+/-- Extract the packing radius of a Delone set. -/
+lemma exists_packing_radius (D : DeloneSet X) :
+    ∃ r > 0, ∀ ⦃x y⦄, x ∈ D.carrier → y ∈ D.carrier → x ≠ y → r ≤ dist x y :=
+  D.uniformlyDiscrete
+
+/-- Distinct points in a Delone set are at positive distance. -/
+lemma dist_pos_of_ne {D : DeloneSet X} {x y : X}
+    (hx : x ∈ D.carrier) (hy : y ∈ D.carrier) (hne : x ≠ y) :
+    0 < dist x y := by
+  obtain ⟨r, hr, hsep⟩ := D.exists_packing_radius
+  exact lt_of_lt_of_le hr <| hsep hx hy hne
+
+/-- At most one point of a Delone set lies in a sufficiently small ball. -/
+lemma subset_ball_singleton (D : DeloneSet X) :
+    ∃ r > 0, ∀ x y z, x ∈ D.carrier → y ∈ D.carrier → z ∈ D.carrier →
+      dist x z < r / 2 → dist z y < r / 2 → x = y := by
+  obtain ⟨ρ, hρ, hsep⟩ := D.exists_packing_radius
+  refine ⟨ρ, hρ, fun x y z hx hy hz hxz hyz ↦ ?_⟩
+  have hxy_lt : dist x y < ρ := by
+    have := calc
+      dist x y ≤ dist x z + dist z y := dist_triangle _ _ _
+      _ < ρ / 2 + ρ / 2 := add_lt_add hxz hyz
+      _ = ρ := by ring
+    simpa using this
+  by_contra hne
+  exact lt_irrefl _ (lt_of_lt_of_le hxy_lt <| hsep hx hy hne)
+
+/-- The image of a Delone set under an isometry is a Delone set. -/
+def map (f : X ≃ᵢ Y) (D : DeloneSet X) : DeloneSet Y := {
+  carrier := f '' D.carrier
```

Comment:
So, if you define this for `UniformSpace`s instead, then you should be able to have this result using `X ≃ᵤ Y`, which is more general. As it stands, the right generality for your current version would of course be bilipschitz maps, but we don't have a bundled version of those.